### ISSUE TITLE
Exploration: export as PDF

### ISSUE THREAD

**Issue by yzhang-gh** [yzhang-gh/vscode-markdown#875]
- [ ] Choose a backend
  - [Puppeteer](https://github.com/puppeteer/puppeteer). Most promising.
    - An issue is it doesn't seem to generate bookmarks for the headings
  - User-installed Chrome. (Not sure whether there is any difference from Puppeteer, e.g. APIs, rendering.)
  - [wkhtmltopdf](https://github.com/wkhtmltopdf/wkhtmltopdf) **✗**
    - A big problem is CJK fonts are not applied.

- [ ] Download (on demand) binaries for different OSes. See [vscode-ltex](https://github.com/valentjn/vscode-ltex) for an example.
  - Handle possible network issues (fallback URLs, proxies etc.)

- [ ] Extra: provide a few presets (themes) @yzhang-gh

cc #270.

**Comment by ComFreek** [comment]
Note that Chrome's print-to-pdf functionality does not always retain working hyperlinks in the output PDF, see here: https://superuser.com/questions/809627/print-webpage-to-pdf-with-working-hyperlinks

**Comment by yzhang-gh** [comment]
As far as I know, it works well in general. (The superuser link looks quite outdated.)

**Comment by dschuessler** [comment]
Are you aware of https://github.com/yzane/vscode-markdown-pdf?

**Comment by yzhang-gh** [comment]
Yep, I know it from the very beginning. It uses Puppeteer now so it also has the problem that it doesn't generate bookmarks.

**Comment by matrix-ranger** [comment]
Why not consider `pandoc`? It works well with bookmarks. Check out this [vscode-pandoc](https://marketplace.visualstudio.com/items?itemName=DougFinke.vscode-pandoc). Note that Typora uses `pandoc` to export to many formats including epub, rtf, LaTeX, docx, etc. but not PDF and HTML. I guess Typora uses the print API of Electron to export PDF? There is an open source markdown editor similar to Typora called [Mark Text](https://github.com/marktext/marktext) which seems to use this API.

Besides, FYI [Markdown Extended](https://marketplace.visualstudio.com/items?itemName=jebbs.markdown-extended#Exporter) uses Puppeteer and `nbconvert` of Jupyter Notebook uses both pandoc and Puppeteer (provides two versions of PDF export).

**Comment by yzhang-gh** [comment]
> `pandoc`

The PDF is generated by TeX so the layout is totally different. If users want that, they already have vscode-pandoc as you mentioned...

> I guess Typora uses the print API of Electron to export PDF? There is an open source markdown editor similar to Typora called [Mark Text](https://github.com/marktext/marktext) which seems to use this API.

Mark Text does use the [API](https://www.electronjs.org/docs/api/web-contents#contentsprinttopdfoptions) from Electron. But I guess it doesn't have bookmarks (unlike Typora).

https://github.com/marktext/marktext/blob/0fa6c33c44e92d0d1729d3d557d02ac7db87e73f/src/main/menu/actions/file.js#L72

**Comment by matrix-ranger** [comment]
I see what you mean. They say nothing about what Typora uses for its PDF export on their website. Maybe can try [WeasyPrint](https://github.com/Kozea/WeasyPrint). This one works well with both bookmarks and CJK.

**Comment by yzhang-gh** [comment]
> WeasyPrint

Interesting. Adding to the list.

**Comment by scott-parkhill** [comment]
I just wanted to say that I am so happy this is under consideration because I love this extension so much, but what annoys me is having to export everything to HTML, then open the HTML in Firefox, then save the page as a PDF, and in many cases then have to use another program to merge the pdfs together into one file. I just wanted to say that I'm so happy this issue is open and I hope it's progressing well! Thank you for your work, this is a great extension!

**Comment by javix64** [comment]
hey @yzhang-gh any update about this nice feature?

**Comment by ruiyuanlu** [comment]
Hi, any updates here?

It will be really help if pdf can be generated.

One more issue about the markdown-pdf extension is the lack of math formula conversion support. 

Although this issue can be fixed by adding following code, but this fix doesnt work in private net. 

'''javascript
<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/x-mathjax-config"> MathJax.Hub.Config({ tex2jax: {inlineMath: [['$', '$']]}, messageStyle: "none" });</script>
'''

Any plan to add support?

**Comment by yzhang-gh** [comment]
Hi, thank both of you for the interest.
As you may already know, (up to now) there is no good API for the PDF related functionality. And different external tools have different issues, any of the solutions not only requires great effort but also is not that satisfactory.
It might be a better idea to make improvements to the existing markdown-pdf extension rather than implementing a new one.

**Comment by javix64** [comment]
Thanks for answering @yzhang-gh, I know what you mean. Thanks for your effort and your work, excellent work!

I guess it is a good idea if you write explicity that this addon does not support PDF (and write which one can do it, if you want) in the README.md section and also is not planned, just for clarify the situation like many people open the issue about this topic.

Have a good developing time!
